core/src/main/java/org/apache/calcite/rel/RelCollationTraitDef.java

```diff
@@ -69,10 +70,14 @@ public RelNode convert(
       return null;
     }
 
-    // Create a logical sort, then ask the planner to convert its remaining
-    // traits (e.g. convert it to an EnumerableSortRel if rel is enumerable
-    // convention)
-    final Sort sort = LogicalSort.create(rel, toCollation, null, null);
+    // Create a sort operator based on given convention
+    Convention toConvention = rel.getConvention();
+    RelBuilder builder =
+        RelFactories.LOGICAL_BUILDER.create(rel.getCluster(), null);
+    RelNode sort = builder.push(rel)
+                          .adoptConvention(toConvention)
+                          .sort(toCollation)
+                          .build();
```

Comment:
Based on the new APIs, isn't simpler to use the following code:
```
RelFactories.SortFactory sortFactory = toConvention.getRelFactories().sortFactory;
Sort sort = (Sort) sortFactory.createSort(rel, toCollation, null, null);
```
It also avoids creating a builder that you will just throw away.

Reply:
This was the very first proposal from Haisheng. We debated a lot and came to agreement to use RelBuilder due to a few benefits - 1. reuse existing RelBuilder logics. For example, it would be very easy to create sort limit as well. 2. support more scenarios, not just the enforcement. 3. support building rel tree with multiple conventions (as discussed in JIRA)

Reply:
I guess I missed again this discussion (and still don't find it :)) apologies for that. Just to be sure we are on the same page I am just talking for this piece of code where we are only doing enforcement. I'm not questioning the other scenarios.

Reply:
The discussion is in JIRA 2970 and also the thread "[DICUSS] Support building physical RelNode in Calcite". There would be other ways to use RelBuilder for physical nodes. I would need to get this in first to work on the subsequent changes. I hesitate to include them in this PR since I believe they are separated.